test/celeritas/grid/SplineXsCalculator.test.cc

```diff
@@ -0,0 +1,446 @@
+        EXPECT_SOFT_EQ(3, calc[0]);
+        EXPECT_SOFT_EQ(3, calc[2]);
+        EXPECT_SOFT_EQ(3, calc[5]);
+
+        // Test between grid points
+        EXPECT_SOFT_EQ(3, calc(Energy{0.2}));
+        EXPECT_SOFT_EQ(3, calc(Energy{5}));
+        EXPECT_SOFT_EQ(3, calc(Energy{2e3}));
+
+        // Test out-of-bounds: cross section still scales according to 1/E
+        // (TODO: this might not be the right behavior for
+        EXPECT_SOFT_EQ(3, calc(Energy{0.0001}));
+        EXPECT_SOFT_EQ(0.3, calc(Energy{1e5}));
+
+        // Test energy grid bounds
+        EXPECT_SOFT_EQ(0.1, value_as<Energy>(calc.energy_min()));
+        EXPECT_SOFT_EQ(1e4, value_as<Energy>(calc.energy_max()));
+    }
+}
+
+TEST_F(SplineXsCalculatorTest, scaled_highest)
+{
+    // values of 1, 10, 100 --> actual xs = {1, 10, 1}
+    this->build(1, 100, 3);
+    this->set_prime_index(2);
+
+    size_type order = 1;
+
+    SplineXsCalculator calc(this->data(), this->values(), order);
+    EXPECT_SOFT_EQ(1, calc(Energy{0.0001}));
+    EXPECT_SOFT_EQ(1, calc(Energy{1}));
+    EXPECT_SOFT_EQ(10, calc(Energy{10}));
+    EXPECT_SOFT_EQ(2.0, calc(Energy{90}));
+
+    // Test access by index
+    EXPECT_SOFT_EQ(1, calc[0]);
+    EXPECT_SOFT_EQ(10, calc[1]);
+    EXPECT_SOFT_EQ(1, calc[2]);
+
+    // Final point and higher are scaled by 1/E
+    EXPECT_SOFT_EQ(1, calc(Energy{100}));
+    EXPECT_SOFT_EQ(.1, calc(Energy{1000}));
+
+    // Test energy grid bounds
+    EXPECT_SOFT_EQ(1, value_as<Energy>(calc.energy_min()));
+    EXPECT_SOFT_EQ(100, value_as<Energy>(calc.energy_max()));
+}
+
+TEST_F(SplineXsCalculatorTest, quadratic_simple)
+{
+    auto reference_xs = [](real_type energy) {
+        auto result = 0.1 * energy * energy;
+        return result;
+    };
+
+    this->build({1e-3, 1e2}, 6, reference_xs);
+
+    for (size_type order = 2; order < 5; ++order)
+    {
+        SplineXsCalculator calc(this->data(), this->values(), order);
+
+        for (real_type e : {1e-2, 5e-2, 1e-1, 5e-1, 1.0, 5.0, 1e1, 5e1, 1e2})
+        {
+            EXPECT_SOFT_EQ(reference_xs(e), calc(Energy{e}));
+        }
+
+        // Test access by index
+        EXPECT_SOFT_EQ(reference_xs(1e-3), calc[0]);
+        EXPECT_SOFT_EQ(reference_xs(1e-1), calc[2]);
+        EXPECT_SOFT_EQ(reference_xs(1e2), calc[5]);
+
+        // Test out-of-bounds
+        EXPECT_SOFT_EQ(reference_xs(1e-3), calc(Energy{1e-5}));
+        EXPECT_SOFT_EQ(reference_xs(1e2), calc(Energy{1e5}));
+
+        // Test energy grid bounds
+        EXPECT_SOFT_EQ(1e-3, value_as<Energy>(calc.energy_min()));
+        EXPECT_SOFT_EQ(1e2, value_as<Energy>(calc.energy_max()));
+    }
+}
+
+TEST_F(SplineXsCalculatorTest, quadratic_scaled_lowest)
+{
+    auto reference_xs = [](real_type energy) {
+        auto result = 0.1 * energy * energy;
+        return result;
+    };
+
+    this->build({1.0e-3, 1e2}, 6, reference_xs);
+    this->convert_to_prime(0);
+    for (size_type order = 2; order < 5; ++order)
+    {
```

Comment:
To see where exactly the single-precision failure is occurring, add
```suggestion
    {
        SCOPED_TRACE("order=" + std::to_string(order));
```

It's probably nothing to worry about, but if there's some surprise numeric instability the single-precision test will usually find it :)

Reply:
Added this to the quadratic tests. I'd guess that the failures are just caused by interpolating values with too large of a magnitude difference for single-precision to not lose accuracy

Reply:
@lebuller I took a more careful look and this was because of accidental promotion to double-precision arithmetic: `0.1` is a double literal, so the reference value was doing the computation at a different precision level leading to slightly different results. (I.e., the single-precision was imprecise by 1e-7 but the reference should've had the same imprecise value.)